### PR TITLE
RAGL S12 updates

### PR DESCRIPTION
**Pull request by suhrm1**
Hi,
I packaged all changes that were implemented during RAGL S12 into this PR.

A couple of newly added features are split across the different commits:

* Added `/games/json` endpoint to retrieve all game results as a JSON array
* Added a spoiler-free display of latest games page
* Fixed a bug where gold/silver medals were not actually displayed based on playoff results
* Added a mechanism that allows to register forfeit games, i.e. player does not drop out of the entire league but some games get awarded to opposition based on administrative decision

Last commit contains all the S12-specific configuration: final player list, map pack, forfeits, playoffs, etc.

Let me know if anything needs to be changed!

### CODE REVIEW

File: raglweb/__init__.py [hunk 1]
```diff
@@ -45,6 +45,12 @@ def _db_close(e=None):
         db.close()
 
 
+def _ordered_alphanumerically(x: str, y: str) -> bool:
+    # returns the alphanumerically higher-ranked string, i.e. "abcde" before "efgh"
+    # NB: always returns a lowercase string!
+    return True if sorted([x.lower(), y.lower()])[0] == x.lower() else False
```
Comment (ubitux) [inline raglweb/__init__.py:52]: `return x.lower() < y.lower()`?

File: raglweb/__init__.py [hunk 2]
```diff
@@ -45,6 +45,12 @@ def _db_close(e=None):
         db.close()
 
 
+def _ordered_alphanumerically(x: str, y: str) -> bool:
+    # returns the alphanumerically higher-ranked string, i.e. "abcde" before "efgh"
```
Comment (ubitux) [inline raglweb/__init__.py:50]: returns *true if*? It's night returning a string

File: raglweb/__init__.py [hunk 3]
```diff
@@ -198,7 +204,21 @@ def playoffs():
 def games():
     db = _db_get()
     games = _get_games(db, 'outcomes')
-    return render_template('games.html', games=games)
+
+    # create an unspoiled list of games (players ordered alphanumerically, additional "winner" attribute)
+    unspoiled_games = []
+    for game in games:
+        g = game.copy()
```
Comment (ubitux) [inline raglweb/__init__.py:222]: A copy can be expensive but fortunately you can avoid it:
```python
    for game in games:
        p0, p0_id, p1, p1_id = game['p0'], game['p0_id'], game['p1'], game['p1_id']
        if not _ordered_alphanumerically(p0, p1):
            game.update(p0=p1, p0_id=p1_id, p1=p0, p1_id=p0)
        game['winner'] = p0
        unspoiled_games.append(g)
```

File: raglweb/__init__.py [hunk 4]
```diff
@@ -21,6 +21,7 @@
 import sqlite3
 from datetime import date, datetime, timedelta
 from .playoffs import get_playoff2, get_playoff4, PlayoffOutcome
+from .forfeit_games import *
```
Comment (ubitux) [inline raglweb/__init__.py:24]: please avoid `import *`, it's hard to keep track of cross references

File: raglweb/__init__.py [hunk 5]
```diff
@@ -109,17 +111,26 @@ def scoreboards():
             status = ''  # TODO: handle late status
 
         rows.append(dict(
-            row_id=len(rows) + 1,
+            # row_id=1
```
Comment (ubitux) [inline raglweb/__init__.py:114]: please just remove the line

File: raglweb/__init__.py [hunk 6]
```diff
@@ -109,17 +111,26 @@ def scoreboards():
...
             status=status or '',
         ))
+
+        # Take into account forfeit games when sorting the list
+        rows = sorted(rows, key = lambda row:(row["wins"], row["winrate"]), reverse=True)
+        # Set player rank based on sorted list
+        i = 1
+        for r in rows:
+            r['row_id'] = 'SF' if "Season Forfeit" in r['status'] else i
+            i += 1
```
Comment (ubitux) [inline raglweb/__init__.py:132]: I don't understand: you aare setting the same key several times?

Btw, you can avoid the `i` counter by doing `for i, r in enumerate(row, 1): ...`

File: raglweb/__init__.py [hunk 7]
```diff
@@ -109,17 +111,26 @@ def scoreboards():
...
-            losses=losses,
-            winrate=wins / nb_played * 100 if nb_played else 0,
+            wins=wins + forfeit_wins,
+            losses=losses + forfeit_losses,
+            winrate=(wins + forfeit_wins) / nb_played * 100 if nb_played else 0,
             status=status or '',
         ))
+
+        # Take into account forfeit games when sorting the list
+        rows = sorted(rows, key = lambda row:(row["wins"], row["winrate"]), reverse=True)
```
Comment (ubitux) [inline raglweb/__init__.py:127]: style: `sorted(rows, key=lambda row: (row["wins"], row["winrate"]), reverse=True)`

File: raglweb/forfeit_games.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+def get_player_forfeit_losses(db: Connection, player_id: int) -> int:
+    res = db.execute(f"SELECT COUNT(*) "
+                     f"FROM forfeit_games "
+                     f"WHERE profile_id1={player_id}").fetchone()[0]
+    return res
+
+
+def get_player_forfeit_games(db: Connection, player_id: int) -> dict:
+    query = f"SELECT fg.profile_id0, (SELECT profile_name FROM players WHERE profile_id=fg.profile_id0) as p0_name, " \
+            f"fg.profile_id1, (SELECT profile_name FROM players WHERE profile_id=fg.profile_id1) as p1_name, " \
```
Comment (ubitux) [inline raglweb/forfeit_games.py:21]: Not a blocking comment nested select sounds pretty expensive. Similarly, running a new query in a loop (the other 2 functions are called like that) is going to be quite a hit on the server

File: raglweb/forfeit_games.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    for row in res:
+        opponent_id = row[0] if row[0] != player_id else row[2]
+        opponent_name = row[1] if row[0] != player_id else row[3]
+        game = {
+            'opponent': opponent_name,
+            'opponent_id': opponent_id,
+            'date': datetime.fromisoformat(row[4]).strftime('%Y-%m-%d'),
+            'map': row[5],
+            'outcome': "Won" if row[0] == player_id else "Lost",
+            'hash': ""
```
Comment (ubitux) [inline raglweb/forfeit_games.py:36]: nit: you may want to avoid mixing double a single quotes

**Review by ubitux — CHANGES_REQUESTED** [review]
Thanks. Can you add a `ragl:` prefix to the commits?

File: raglweb/__init__.py (hunk 1 above)
Comment (suhrm1) [reply raglweb/__init__.py:52]: Thanks for the quick feedback, will have a look at your remarks.

I wonder what was the case that dragged me into this specific rabbit hole though. The comments are from some earlier attempt and I wonder how I did not end up with the simple solution you proposed. :facepalm: :shrug:

File: raglweb/__init__.py (hunk 6 above)
Comment (suhrm1) [reply raglweb/__init__.py:132]: The original ranking does not take into account the wins/losses from forfeit games. I re-order the list after adding in those wins and then explicitly add the "row_id" attribute which is only used to display the ranking in the scoreboard table. Players that forfeit the entire season should not be actually ranked but just display "SF".

Does that clarify it?

File: raglweb/forfeit_games.py (hunk 8 above)
Comment (suhrm1) [reply raglweb/forfeit_games.py:26]: I replaced the 2 other functions you mention with one that just grabs forfeit games win/loss stats for all players. Left the nested query as is though for lack of good ideas on how to change it.

File: raglweb/__init__.py (hunk 6 above)
Comment (ubitux) [reply raglweb/__init__.py]: Yes, thanks

File: laddertools/ragl.py [hunk 10]
```diff
@@ -218,6 +219,10 @@ def _main(args):
     if playoffs:
         _handle_extra_outcomes(c, extra_outcomes, playoffs)
 
+    if 'Forfeit_Games' in players_info.keys():
+        forfeit_games = [game for game in players_info['Forfeit_Games']]
```
Comment (ubitux) [inline laddertools/ragl.py:223]: Why can't you use `players_info['Forfeit_Games']` directly? If it really needs to be a list you can use `list(players_info['Forfeit_Games'])` but I'm not sure you need that.

File: raglweb/__init__.py [hunk 11]
```diff
@@ -198,7 +204,18 @@ def playoffs():
...
+
+    # create an unspoiled list of games (players ordered alphanumerically, additional "winner" attribute)
+    unspoiled_games = []
+    for game in games:
+        p0, p0_id, p1, p1_id = game['p0'], game['p0_id'], game['p1'], game['p1_id']
+        # re-order players alphanumerically if necessary
+        if not _ordered_alphanumerically(p0, p1):
+            game.update(p0=p1, p0_id=p1_id, p1=p0, p1_id=p0_id)
+        game['winner'] = p0
+        unspoiled_games.append(game)
```
Comment (ubitux) [inline raglweb/__init__.py:230]: Now that I think again, do you even need the `unspoiled_games` array? I believe the loop is mutating the values of `games` in the loop anyway so it shouldn't be needed.

**Review by ubitux — CHANGES_REQUESTED** [review]
Should be good after my last two comments are honored.

I'm too lazy to test tbh so I'll trust your changes, so make sure to test your last iteration.

Thanks

File: laddertools/ragl.py (hunk 10 above)
Comment (suhrm1) [reply laddertools/ragl.py]: Yep, that worked.

File: raglweb/__init__.py (hunk 11 above)
Comment (suhrm1) [reply raglweb/__init__.py]: This also worked, well spotted :+1:

**Comment by suhrm1** [comment]
Latest version is now running at ragl.org and seems to work as expected

**Review by ubitux — APPROVED** [review]

**Pull request merged**